File: server/tables/keyword_usage.py
from app import db

class KeywordUsage(db.Model):
    __tablename__ = 'keyword_usage'
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    transcript_id = db.Column(db.Integer, db.ForeignKey('transcript.id'), nullable=False)
    word = db.Column(db.String(64), nullable=False)
    keyword = db.Column(db.String(64), nullable=False)
    similarity = db.Column(db.Float, nullable=False)

    WORD_MAX_LENGTH = 64
    KEYWORD_MAX_LENGTH = 64

    def __hash__(self):
        return hash((self.id))

    def __init__(self, transcript_id, word, keyword, similarity):
        self.transcript_id = transcript_id
        self.word = word
        self.keyword = keyword
        self.similarity = similarity

    def json(self, suppress=[]):
        json_data = dict(
            id=self.id,
            word=self.word,
            keyword=self.keyword,
            similarity=self.similarity
        )
        if not 'transcript_id' in suppress:
            json_data['transcript_id']=self.transcript_id
        return json_data

    @staticmethod
    def verify_fields(word=None, keyword=None):
        message = None
        if word != None:
            if len(word) > KeywordUsage.WORD_MAX_LENGTH:
                message = 'Word must not exceed {0} characters.'.format(KeywordUsage.WORD_MAX_LENGTH)
        if keyword != None:
            if len(keyword) > KeywordUsage.KEYWORD_MAX_LENGTH:
                message = 'Keyword must not exceed {0} characters.'.format(KeywordUsage.KEYWORD_MAX_LENGTH)
        return message == None, message

File: frontend/src/models/keyword-usage.js
export class KeywordUsageModel {
  // Server Fields
  id;
  transcript_id;
  word;
  keyword;
  similarity;

  static fromJson(json){
    const model = new KeywordUsageModel();
    model.id = json['id'];
    model.transcript_id = json['transcript_id'];
    model.word = json['word'];
    model.keyword = json['keyword'];
    model.similarity = +((1.0 - json['similarity']) * 100).toFixed(2);
    return model;
  }

  // Converts JSON to KeywordUsageModel[]
  static fromJsonList(jsonArray){
    const keywordUsages = [];
    for (const el of jsonArray) {
      keywordUsages.push(KeywordUsageModel.fromJson(el));
    }
    return keywordUsages;
  }
}
